bpf/lib/encrypt.h

```diff
@@ -205,6 +206,91 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
 #endif /* ENABLE_ROUTING */
 }
 
+static __always_inline int
+ipsec_maybe_redirect_to_encrypt(struct __ctx_buff *ctx, __be16 proto)
+{
+	struct remote_endpoint_info *dst = NULL;
+	struct remote_endpoint_info __maybe_unused *src = NULL;
+	void *data __maybe_unused, *data_end __maybe_unused;
+	struct iphdr __maybe_unused *ip4;
+	__u32 magic __maybe_unused = 0;
+	int ret = 0;
+	__u8 dst_mac = 0;
+
+	if (!eth_is_supported_ethertype(proto))
+		return DROP_UNSUPPORTED_L2;
+
+	switch (proto) {
+# ifdef ENABLE_IPV4
+	case bpf_htons(ETH_P_IP):
+		if (!revalidate_data(ctx, &data, &data_end, &ip4))
+			return DROP_INVALID;
+
+#  if defined(TUNNEL_MODE)
+		/* In tunneling mode IPsec needs to encrypt tunnel traffic,
+		 * so that src sec ID can be transferred.
+		 *
+		 * This also handles IPv6, as IPv6 pkts are encapsulated w/
+		 * IPv4 tunneling.
+		 */
+		if (ctx_is_overlay(ctx)) {
+			ret = set_ipsec_encrypt(ctx, 0, ip4->daddr,
+						get_identity(ctx), false,
+						true);
+			if (ret != CTX_ACT_OK)
+				return ret;
+			goto overlay_encrypt;
+		}
+#  endif /* TUNNEL_MODE */
+
+		dst = lookup_ip4_remote_endpoint(ip4->daddr, 0);
+		src = lookup_ip4_remote_endpoint(ip4->saddr, 0);
+		break;
+# endif /* ENABLE_IPV4 */
+
+# ifdef ENABLE_IPV6
+	case bpf_htons(ETH_P_IPV6):
+		/* ipv6 traffic not currently supported unless its tunneled */
+# endif /* ENABLE_IPv6 */
+	default:
+		return CTX_ACT_OK;
+	}
+
+	/* only hairpin pod-to-pod or overlay traffic back into the stack for
+	 * XFRM encryption
+	 */
+	if (!src || src->sec_identity == HOST_ID)
+		return CTX_ACT_OK;
+	if (!identity_is_cluster(src->sec_identity))
+		return CTX_ACT_OK;
+	if (identity_is_remote_node(src->sec_identity))
+		return CTX_ACT_OK;
+
+	/* mark packet for encryption based on ipcache's key */
+	ret = set_ipsec_encrypt(ctx, dst->key, ip4->daddr, src->sec_identity, false, true);
+	if (ret != CTX_ACT_OK)
+		return ret;
+
+#  if defined(TUNNEL_MODE) && defined(ENABLE_IPV4)
+overlay_encrypt:
+#  endif
+	/*
+	 * source mac is our current egress interface, lets copy it to dmac
+	 * so redirecting to ingress side of the same interface doesn't fail.
+	 */
+	if (eth_load_saddr(ctx, &dst_mac, 0) != 0)
+		return DROP_INVALID;
+	if (eth_store_daddr(ctx, &dst_mac, 0) != 0)
+		return DROP_WRITE_ERROR;
+
+	/* redirect to ingress side of ifindex so the packet has xfrm applied */
+	ret = ctx_redirect(ctx, ctx->ifindex, BPF_F_INGRESS);
+	if (ret != CTX_ACT_REDIRECT)
+		return DROP_INVALID;
```

Comment:
I wonder if we can/should route through `to_host@cilium_host` instead, similar to what `ctx_redirect_to_proxy_host_egress()` is doing. Would that allow us to avoid the sysctl changes?

Reply:
`accept_local` is already set on the `cilium_host` / `cilium_net` interfaces. See the `enableForwarding()` calls in https://github.com/cilium/cilium/blob/4c2a11690660e1e274d362b46ae95a40cf5a82ba/pkg/datapath/loader/netlink.go#L114

Reply:
Yeah I still remember that Joe suggested the same thing months ago when reviewing multicasting IPsec.

I didn't have a strong opinion before, but now that I understand cilium has an existing hairpin pattern (redirect from native dev to cilium_net egress), plus we don't have to set additional sysctl, so let's try.

Reply:
Yup, going to evaluate this once I get my head out of this connectivity test mess I'm in. 

My only hesitations are traversing the veth pair and not spreading this IPSec logic across more eBPF programs.
For instance, lets say we get all the way to the egress device. We can mark the IPsec packet there. 
Then, we redirect to `cilium_net` egress. We now need to look for packets marked for encryption there and let them pass thru to the ingress side of `cilium_host`, at which point we need to do the same thing, until we get to the stack. We go through a lot more code this way. 

There is the idea of just redirect to ingress side of `cilium_net`, in which case I don't think we'll traverse the veth pair as we'll look like we are coming ingress into the kernel. But this is not what's being suggested in that function pointed to.

Reply:
@julianwiedmann @jschwinger233 

I explored the simpler route of just redirect to `ingress@cilium_net`. This seems to work well and removes the requirement for the sysctl changes. 

This diff works fine:

```diff
diff --git a/bpf/lib/encrypt.h b/bpf/lib/encrypt.h
index 4fad781bb5..12b5b7e69a 100644
--- a/bpf/lib/encrypt.h
+++ b/bpf/lib/encrypt.h
@@ -233,7 +233,7 @@ ipsec_maybe_redirect_to_encrypt(struct __ctx_buff *ctx, __be16 proto)
        struct ipv6hdr __maybe_unused *ip6;
        __u32 magic __maybe_unused = 0;
        int ret = 0;
-       __u8 dst_mac = 0;
+       union macaddr dst_mac = CILIUM_NET_MAC;

        if (!eth_is_supported_ethertype(proto))
                return DROP_UNSUPPORTED_L2;
@@ -304,20 +304,19 @@ ipsec_maybe_redirect_to_encrypt(struct __ctx_buff *ctx, __be16 proto)
 #  if defined(TUNNEL_MODE) && defined(ENABLE_IPV4)
 overlay_encrypt:
 #  endif
-       /*
-        * source mac is our current egress interface, lets copy it to dmac
-        * so redirecting to ingress side of the same interface doesn't fail.
+       /* redirect to the ingress side of CILIUM_NET.
+        * this will subject the packet to the ingress XFRM hooks,
+        * encrypting the packet.
+        *
+        * the encrypted packet will be recirculated to the stack and the final
+        * egress will occur toward the IPsec tunnel's destination.
         */
-       if (eth_load_saddr(ctx, &dst_mac, 0) != 0)
-               return DROP_INVALID;
-       if (eth_store_daddr(ctx, &dst_mac, 0) != 0)
+       if (eth_store_daddr(ctx, (const __u8 *)&dst_mac, 0) != 0)
                return DROP_WRITE_ERROR;

-       /* redirect to ingress side of ifindex so the packet has xfrm applied */
-       ret = ctx_redirect(ctx, ctx->ifindex, BPF_F_INGRESS);
+       ret = ctx_redirect(ctx, HOST_IFINDEX, BPF_F_INGRESS);
        if (ret != CTX_ACT_REDIRECT)
                return DROP_INVALID;
-
        return ret;
 }
```

If the major goal here is to remove the sysctl from native devices, this will accomplish it. I'm happy to make this change. 

I don't quite see the need to traverse the VETH pair directly, as we don't gain any benefit that I can see from going thru both veth pairs.

Reply:
> I don't quite see the need to traverse the VETH pair directly, as we don't gain any benefit that I can see from going thru both veth pairs.

Agree, I never understood why `ctx_redirect_to_proxy_hairpin_ipv4()` redirects packets to cilium_net egress instead of cilium_host ingress directly.

Reply:
Latest push adds the change and also removes the sysctl related commit as its no longer necessary.